stl/inc/xstring

```diff
@@ -3616,6 +3616,25 @@ public:
         return _Mypair._Myval2._Myres;
     }
 
+#if _HAS_CXX20
+    void reserve(_CRT_GUARDOVERFLOW const size_type _Newcap) { // determine new minimum length of allocated storage
+        if (_Mypair._Myval2._Myres < _Newcap) { // reallocate to grow
+            const size_type _Old_size = _Mypair._Myval2._Mysize;
+            _Reallocate_grow_by(
+                _Newcap - _Old_size, [](_Elem* const _New_ptr, const _Elem* const _Old_ptr, const size_type _Old_size) {
+                    _Traits::copy(_New_ptr, _Old_ptr, _Old_size + 1);
+                });
+
+            _Mypair._Myval2._Mysize = _Old_size;
+            return;
+        }
+
+        // ignore requests to reserve to [0, _Myres]
+    }
+
+
+    _CXX20_DEPRECATE_STRING_RESERVE_DEFAULT_ARGUMENT void reserve() {} // no-op

```

Comment:
I don't think it's a good idea to change the behavior of `reserve`-with-no-explicit-argument in 20 mode. It's deprecated, sure - that's fine - but I see no reason we should change the behavior of programs here simply because they flip the switch from C++17 to C++20.

Reply:
When I looked at the current implementation of zero-arg `reserve()`, I thought it was a no-op, and told Nathan so:

https://github.com/microsoft/STL/blob/53cdb9f8a85c2500077ccf0676619110340c03bf/stl/inc/xstring#L3619-L3646

I was wrong!

The initial check makes this a no-op for non-empty strings:

https://github.com/microsoft/STL/blob/53cdb9f8a85c2500077ccf0676619110340c03bf/stl/inc/xstring#L3620-L3622

However, for empty strings, we get to:

https://github.com/microsoft/STL/blob/53cdb9f8a85c2500077ccf0676619110340c03bf/stl/inc/xstring#L3639-L3643

and if we're in large mode (i.e. holding a dynamically allocated buffer), we deallocate it.

So this is `shrink_to_fit()` only when `empty()`. Example:

```
C:\Temp>type meow.cpp
#include <iostream>
#include <string>
using namespace std;

void test(string& s) {
    cout << "string: \"" << s << "\"\n";
    cout << "before: size " << s.size() << ", ";
    cout << "capacity " << s.capacity() << "\n";
    s.reserve();
    cout << " after: size " << s.size() << ", ";
    cout << "capacity " << s.capacity() << "\n\n";
}

int main() {
    string s1("This is a very long string.");
    s1.append(1000, '-');
    s1.erase(40);
    test(s1);

    string s2("short");
    test(s2);

    s1.clear();
    test(s1);

    s2.clear();
    test(s2);
}

C:\Temp>cl /EHsc /nologo /W4 /MTd meow.cpp && meow
meow.cpp
string: "This is a very long string.-------------"
before: size 40, capacity 1039
 after: size 40, capacity 1039

string: "short"
before: size 5, capacity 15
 after: size 5, capacity 15

string: ""
before: size 0, capacity 1039
 after: size 0, capacity 15

string: ""
before: size 0, capacity 15
 after: size 0, capacity 15
```

As far as I can tell, implementing `reserve()` with `if (empty()) { shrink_to_fit(); }` would leave the behavior unchanged, unless I am wrong yet again.

Reply:
> As far as I can tell, implementing `reserve()` with `if (empty()) { shrink_to_fit(); }` would leave the behavior unchanged, unless I am wrong yet again.

That's my reading as well. I suppose we'll need @BillyONeal or @barcharcraz to tell us that we're both wrong.

Reply:
This also means that I should add a regression test to verify this behavior when we check this in.

Reply:
Thanks for the help!
One more question though.
The convention is to use
`if (_Mypair._Myval2._Mysize == 0)` over `if (empty())` correct?

Reply:
Second followup:
In the case where some calls `reserve(x)` where `_BUFFER_SIZE > x` should the string `_Become_small()`?
In this case, should the old implementation of reserve() should be?

Reply:
We typically prefer the former for smaller debug codegen, but `reserve`-with-no-arguments is "weird" enough that I wouldn't complain either way.